### PR TITLE
i#2433: Implement opnd_compute_address for AArch64

### PR DESCRIPTION
**Pull request by yury-khrustalev**
Fixes: #2433 Implement use of extension and shift in memory reference operands for AArch64.

### CODE REVIEW

**Comment by yury-khrustalev** [comment]
I'm not sure how the failure of `code_api|client.alloc-noreset` is related to this change.

**Comment by yury-khrustalev** [comment]
Can someone please restart tests? Thank you!

**Comment by johnfxgalea** [comment]
run arm tests

**Comment by johnfxgalea** [comment]
"run arm tests" command seems to have stalled for me.

@AssadHashmi Any idea what is wrong (at your convenience) please.

**Comment by AssadHashmi** [comment]
run arm tests

**Comment by AssadHashmi** [comment]
> "run arm tests" command seems to have stalled for me.
> 
> @AssadHashmi Any idea what is wrong (at your convenience) please.

Something appears to have changed on the GitHub side of the trigger interface, as our Jenkins CI has not changed for months:

**Comment by AssadHashmi** [comment]
> @AssadHashmi Any idea what is wrong (at your convenience) please.

Something appears to have changed on the GitHub side of the trigger interface or we've reached some sort of limit:
```Nov 16, 2020 12:18:34 PM org.jenkinsci.plugins.ghprb.GhprbRepository initGhRepository
SEVERE: Error while accessing rate limit API
org.kohsuke.github.HttpException: {"message":"Bad credentials. The API can't be accessed using username/password authentication. Please create a personal access token to access t
his endpoint: http://github.com/settings/tokens","documentation_url":"https://docs.github.com/articles/creating-a-personal-access-token-for-the-command-line"}
        at org.kohsuke.github.Requester.handleApiError(Requester.java:703)
        at org.kohsuke.github.Requester._to(Requester.java:306)
```
I'm on triage duty this week and will work on fixing...

**Comment by AssadHashmi** [comment]
Update: turns out GitHub's authentication interface has changed and the old version was deprecated on November 13, 2020 at 16:00 UTC, https://developer.github.com/changes/2020-02-14-deprecating-password-auth/

Requires re-configuring from the Jenkins side...

**Comment by drbot0** [comment]
Test message from Jenkins CI using Personal Access Token

**Comment by yury-khrustalev** [comment]
Can we also restart Travis CI tests? I think the current failure is flaky and not related to the patch (this test was previously passing for the patch).

**Comment by AssadHashmi** [comment]
> Can we also restart Travis CI tests? I think the current failure is flaky and not related to the patch (this build was previously passing for the patch).

I've re-started the build. You should be able to click on the 'Details' link, sign-in using your GitHub account and re-start.
Let us know if you can't.

**Comment by johnfxgalea** [comment]
@AssadHashmi Thanks for fixing ARM tests so fast.

**Comment by derekbruening** [comment]
The commit description is missing the link (either `Issue:` or `Fixes`): #2433

**Comment by yury-khrustalev** [comment]
Looks like one of the test builds didn't complete. Could you please restart Travis CI jobs? Thank you!

**Comment by yury-khrustalev** [comment]
Could you please restart tests? Thank you!

**Comment by AssadHashmi** [comment]
Hi @derekbruening @abhinav92003, any thoughts on why `tool.drcachesim.delay-simple` could be failing in this patch?
http://139.178.83.194:8080/job/DynamoRIO-AArch64-Precommit/1884/console

The following assert is triggered:
```
<Application /var/lib/jenkins/.jenkins/workspace/DynamoRIO-AArch64-Precommit/build/build_debug-internal-64/suite/tests/bin/simple_app (32711).  Internal Error: DynamoRIO debug check failure: /var/lib/jenkins/.jenkins/workspace/DynamoRIO-AArch64-Precommit/core/heap.c:1961 IF_WINDOWS(doing_detach ||) vmh->num_free_blocks == vmh->num_blocks - unfreed_blocks || ((ever_beyond_vmm IF_WINDOWS(|| get_os_version() >= WINDOWS_VERSION_8_1)) && vmh->num_free_blocks >= vmh->num_blocks - unfreed_blocks)
(Error occurred @862 frags)                                               
version 8.0.18590, custom build
```

Also, I notice that the migration to GitHub Actions is still in progress.
Should we wait until the new CI is stable before continuing work on this?

**Comment by yury-khrustalev** [comment]
The diff between target upstream branch and PR branch remains the same throughout all updates from upstream. Precommit and other tests fail randomly and errors don't seem to be related to the patch. On several occasions tests were passing successfully.

**Comment by AssadHashmi** [comment]
> Precommit and other tests fail randomly and errors don't seem to be related to the patch.

The AArch64 precommit fails **consistently** on `tool.drcachesim.delay-simple` and only that.

The last time the AArch64 precommit passed for this patch was 24-Nov-2020 07:47:08:
http://139.178.83.194:8080/job/DynamoRIO-AArch64-Precommit/1877/

The first `tool.drcachesim.delay-simple` failure occurred on 24-Nov-2020 16:24:30 for a CI patch:
https://github.com/DynamoRIO/dynamorio/pull/4559 `i4549 Github Action: Add CI test for clang-format, osx and a64-on-x86`

http://139.178.83.194:8080/job/DynamoRIO-AArch64-Precommit/1878/

**Comment by derekbruening** [comment]
Please, for any test failure, list the test name and the issue number that covers the failure here in the PR.  If there is no issue, please file an issue.  Ideally, mark flaky tests in runsuite_wrapper to help keep the CI green (or even better help to fix them).  The only flaky tests I recall seeing in recent PR's were the client.winxfer (#4058) and client.alloc-noreset (i#4436) on Appveyor -- is that what is being hit here?  Someone should put them on the ignore list.

**Comment by derekbruening** [comment]
> Hi @derekbruening @abhinav92003, any thoughts on why `tool.drcachesim.delay-simple` could be failing in this patch?
> http://139.178.83.194:8080/job/DynamoRIO-AArch64-Precommit/1884/console
> 
> The following assert is triggered:
> 
> ```
> <Application /var/lib/jenkins/.jenkins/workspace/DynamoRIO-AArch64-Precommit/build/build_debug-internal-64/suite/tests/bin/simple_app (32711).  Internal Error: DynamoRIO debug check failure: /var/lib/jenkins/.jenkins/workspace/DynamoRIO-AArch64-Precommit/core/heap.c:1961 IF_WINDOWS(doing_detach ||) vmh->num_free_blocks == vmh->num_blocks - unfreed_blocks || ((ever_beyond_vmm IF_WINDOWS(|| get_os_version() >= WINDOWS_VERSION_8_1)) && vmh->num_free_blocks >= vmh->num_blocks - unfreed_blocks)
> (Error occurred @862 frags)                                               
> version 8.0.18590, custom build
> ```

That's a memory leak.  Not sure why this PR would trigger it.  Can you reproduce it locally (run 1000x or sthg) in head or this PR?

> Also, I notice that the migration to GitHub Actions is still in progress.
> Should we wait until the new CI is stable before continuing work on this?

That never affected the a64 Jenkins and the Travis replacements are all in place at this point.

**Comment by derekbruening** [comment]
I filed an issue on delay-simple: #4571

**Comment by derekbruening** [comment]
The delay-simple test failure was caused by #4508 -- @gregcawthorne has the fix but it is not yet committed

**Comment by yury-khrustalev** [comment]
The diff remains the same after recent update of the pull request branch from upstream. So, I'm not sure what can be causing these x86 failures.

**Comment by yury-khrustalev** [comment]
Now all but AArch64 Precommit are failing with no apparent reason in the patch. @derekbruening, could you please have a look? Thank you!

**Comment by derekbruening** [comment]
https://github.com/DynamoRIO/dynamorio/issues/4370#issuecomment-738172341

**Comment by yury-khrustalev** [comment]
Could someone please review the patch? Thank you!

**Comment by derekbruening** [comment]
@AssadHashmi has been listed as the reviewer this whole time

File: core/ir/opnd_shared.c [hunk 1]
```diff
@@ -2119,9 +2119,30 @@ opnd_compute_address_priv(opnd_t opnd, priv_mcontext_t *mc)
...
+        uint amount = 0;
+        dr_extend_type_t type = opnd_get_index_extend(opnd, &scaled, &amount);
         reg_t index_val = reg_get_value_priv(index, mc);
-        /* FIXME i#1569: Compute extension and shift. */
-        scaled_index = index_val;
+        reg_t extended = 0;
+        uint msb = 0;
+        switch (type) {
+        default: CLIENT_ASSERT(false, "Unsupported extend type"); return NULL;
+        case DR_EXTEND_UXTW: extended = (index_val << (63u - 31u)) >> (63u - 31u); break;
```
[the remark this answers is not in the document]
Comment (yury-khrustalev) [reply core/ir/opnd_shared.c:2130]: This is `UXTW`, so it is zero-extension. I thought using `63u - 31u` might make it more readable than just `32u`, and compiler would optimise it anyway.

File: core/ir/opnd_shared.c [hunk 2]
```diff
@@ -2119,9 +2119,30 @@ opnd_compute_address_priv(opnd_t opnd, priv_mcontext_t *mc)
...
+        case DR_EXTEND_SXTW:
+            extended = (index_val << (63u - 31u)) >> (63u - 31u);
+            msb = extended >> 31u;
+            if (msb == 1) {
+                extended = ((~0ull) << 32u) | extended;
```
[the remark this answers is not in the document]
Comment (yury-khrustalev) [reply core/ir/opnd_shared.c:2135]: This is sign extension of `index_val<31:0>`.

**Review by AssadHashmi — APPROVED** [review]

**Pull request merged**